README.md
# modBIN_withXML
modify bin with location recorded by xml

## Work flow :
1. Can change target.txt to modify more step
   1. Default BiosStartInFullRomImage + PcdFlashNvStorageVariableOffset
   2. Rule : one line of feature-name and one line of act 
   3. The act only suppout +/-
   4. If act isn't correct or feature-name cannot find than it will be skiped

2. Find feature-name which user set in {Platform}.xml
   1. Now can find feature-name in tag of first layer and its text is the value we want
   2. Now can find feature-name in value of "name" tag of forth or fifth layers and the value of the tag "value" of the same layer is what we want
   3. User can choice number or press enter to use .xml witch first finded

3. Count the value and print it to double check

4. Read write_data.bin which will replace 
   1. User can input manually to modify when execution CT/MAC/ForceNetBoot/counter/timer otherwise default 11002200...EE00/FF..FF/1/5/2
   2. User can also modify by command line use parameters
   3. Format for 4-1: \x00\x00\..\x07 in one line
   4. Format for 4-2: number of non-spaced strings from 0-9 and A-F
   5. If use arguments -y it will automatically confirm.
   6. Please note that the BIN file will be modified if the arguments y and the four modification arguments m, s, c, and p are used together

5. Write date to {Platform_version}.bin
   1. Copy {Platform_version}.bin and rename {Platform_version}_ori.bin then progrm will modify in {Platform_version}.bin

## Installation
1. Clone the repository: ```git clone https://github.com/yishawnpeng/modBIN_withXML.git```
2. Install Python 3.x or later: https://www.python.org/downloads/
3. Install required libraries: ```pip install -r requirements.txt```

## History
For more history information, please reference /history/HISTORY.txt

## How to use
1. Modify target.txt if you want to change the location of the bin in the calculation result of .xml or just keep default
2. Modify write_data_default.bin or keep default
3. Put your .xml and .bin in the same folder
4. Run ```modBIN_withXML.py``` or ```M1Inject.exe``` in cmd window and attach the arguments you want to use.
5. Input the corresponding information based on the program's feedback.

> Note: Use ```-h``` if you need more infomateion for arguments 

## Contributing
If you would like to contribute to this project, please follow these steps:
 1. Fork the repository
 2. Create a new branch for your feature: ```git checkout -b feature-name```
 3. Make changes and commit them: ```git commit -am 'Add some feature'```
 4. Push to the branch: ```git push origin feature-name```
 5. Submit a pull request

history/history.txt
v4 2023-08-15
-------------
1.Fix bug : The format of SBCT of manual input is not correct.
2.Optimize flow : Add arguments -y to auto confirm all.

v3 2023-08-09
-------------
1.Fix bug : Counter/period input string swap error with function 4-2
2.Fix bug : Can't modify counter/period using command-line-parameters
3.Fix bug : Can't use write_data_factory_default.bin
4.Optimize flow : If only one write_data_xxxx.bin it can skip choose
5.Optimize flow : If use command-line-parameters it will skip the user input to modify binary

v2 2023-08-08
-------------
1.Add function : 4-2 user can input manually to modify when execution
2.Add function : 4-3 user can also modify by command line use parameters
3.Optimize flow : Divide different function into different .py to make the process more streamlined

v1 2023-08-01
-------------
1.Base function

argParse.py
import argparse

version = "4"
macHelpStr = "Input MAC you want. Enter 12 non-spaced strings from 0-9 and A-F."
sbctHelpStr = "Input SystemBordCT you want. Enter 30 non-spaced strings from 0-9 and A-F."
counterHelpStr = "Input Counter you want. Enter 4 non-spaced strings from 0-9 and A-F."
periodHelpStr = counterHelpStr
autoHelpStr = "Automatically confirm all actions without requiring manual confirmation. \
Please note that the .bin file will be modified if the argument y and the four modification parameters m, s, c, and p are used together."

cheeckRange =["0","1","2","3","4","5","6","7","8","9",\
              "a","b","c","d","e","f","A","B","C","D","E","F"]

def checkAll(string, requestLen) :
    if len(string) != requestLen :
        return False
    for i in string :
        if i not in cheeckRange :
            return False
    return True

def argParse_contant(argumentDict):
    exitFlag = False
    autoUpdate = False
    parser = argparse.ArgumentParser(prog='modbin_withxml.py', description='Tutorial')
    parser.add_argument("-v", "--version", action="version", version=version)
    parser.add_argument("-y", "--yes", action="store_true",help=autoHelpStr )
    parser.add_argument("-m", "--mac", help=macHelpStr )
    parser.add_argument("-s", "--sbct", help=sbctHelpStr )
    parser.add_argument("-c", "--counter", help=counterHelpStr )
    parser.add_argument("-p", "--period", help=periodHelpStr )
    args = parser.parse_args()
    if args.yes :
        autoUpdate = True

    if args.mac:
        argumentDict["MAC"] = args.mac
        if not checkAll(args.mac, 12) :
            print("Input MAC format illegal !")
            exitFlag = True
        
    if args.sbct:
        argumentDict["SBCT"] = "".join([args.sbct[i:i+2]+"00"for i in range(0,len(args.sbct),2)])
        if not checkAll(args.sbct, 30) :
            print("Input SystemBordCT format illegal !")
            exitFlag = True
        
    if args.counter:
        argumentDict["Counter"] = args.counter[2:]+args.counter[0:2]
        if not checkAll(args.counter, 4) :
            print("Input Counter format illegal !")
            exitFlag = True
       
    if args.period:
        argumentDict["Period"] = args.period[2:]+args.period[0:2]
        if not checkAll(args.period, 4) :
            print("Input Period format illegal !")
            exitFlag = True

    if exitFlag :
        return autoUpdate, False
    return autoUpdate, True

choceFile.py
import os                                   # operate file
import re                                   # find file

def ChoceFile(file_type, autoUpdate) : # type : xml / bin
    returnName = None
    fatherDir = os.getcwd()
    allDir = os.listdir(fatherDir)
    if file_type == "bin" :
        fileNameList = re.compile("[^write].*\." + file_type)
    elif file_type == "write_data_default.bin" :
        fileNameList = re.compile(file_type)
    else :
        fileNameList = re.compile(".*\." + file_type)
    fileNameList = list( filter( fileNameList.match, allDir ) )

    if len(fileNameList) >= 2 and not autoUpdate : 
        fileNameList = ["("+str(i+1)+")"+fileNameList[i] for i in range(len(fileNameList))]
        fileNameString = ""
        for i in fileNameList : fileNameString += i + " "
        fileNameString = "Input which NUMBER ." \
                            + file_type \
                            + " you want " \
                            + fileNameString \
                            + "\nor press enter to use " \
                            + fileNameList[0][3:] \
                            + " : "
        fileNameNum = input( fileNameString ) or 1
        while not 0 < int(fileNameNum) <= len(fileNameList) :
            print("Not legal input !!!")
            fileNameNum = input( fileNameString ) or 1
        fileName = fileNameList[int(fileNameNum)-1]
        returnName = fileName[3:]
    elif len(fileNameList) == 1 or autoUpdate :
        returnName = fileNameList[0]
    else :
        if file_type == "write_data_default.bin" :
            print("Can not find write_data_default.bin !!!" )
        print("Can not find any ." + file_type + " !!!" )
    
    return returnName

modifyBin.py
from argParse import *          # for userInputStr... and 

binAreaLen = {"startOffset":204,    "SBCT":30,           "MAC":6,\
              "ForceNetBoot":1,     "ForceBackLight":1,  "DisableLidSwitch":1, \
              "Counter":2,          "Period":2,          "endOffset":15 }

sbctEnd = binAreaLen["startOffset"]+binAreaLen["SBCT"]
counterStart = binAreaLen["startOffset"]+binAreaLen["SBCT"]\
                +binAreaLen["MAC"]+binAreaLen["ForceNetBoot"] \
                +binAreaLen["ForceBackLight"]+binAreaLen["DisableLidSwitch"]
counterEnd = counterStart+binAreaLen["Counter"]

binAreaStart = {"SBCT":binAreaLen["startOffset"],\
                "MAC":sbctEnd,\
                "Counter":counterStart,\
                "Period":counterEnd,\
                }

argumentDict={"MAC":None,"SBCT":None,"Counter":None,"Period":None}

userInputData = ["("+str(i)+")"+j for i,j in zip(range(len(argumentDict)),argumentDict)]
userInputStr = "Which number you want to modify \n"+ "\n".join(userInputData) \
                + "\n(Enter) to exit : "

def PrintWriteInfoSize(WriteDataFileName) :
    print("In "+WriteDataFileName+" :")
    print("\tSBCT start at : "+str(binAreaLen["startOffset"]))
    print("\t     end   at : "+str(binAreaLen["startOffset"]+binAreaLen["SBCT"]))
    print("\tMAC start at : "+str(sbctEnd))
    print("\t    end   at : "+str(sbctEnd+binAreaLen["MAC"]))
    print("\tCounter start at : "+str(counterStart))
    print("\t        end   at : "+str(counterStart+binAreaLen["Counter"]))
    print("\tPeriod start at : "+str(counterEnd))
    print("\t       end   at : "+str(counterEnd+binAreaLen["Period"]))


def ModifyWriteBin_manul(WriteDataFileName, argumentDict) :
    reallyMod = input("Input any key if you want to modify data or only enter to skip : ") 
    if reallyMod != "" :
        whichOne = "\n"
        while whichOne :
            whichOne = input(userInputStr) 
            if whichOne == "" :
                break
            elif whichOne == "0" :
                temp = input(macHelpStr[:-1]+" : ")
                if checkAll(temp, 12) :
                    argumentDict["MAC"] = temp
                else : print("Input MAC format illegal !\n")
            elif whichOne == "1" :
                temp = input(sbctHelpStr[:-1]+" : ")
                if checkAll(temp, 30) : 
                    argumentDict["SBCT"] = "".join([temp[i:i+2]+"00"for i in range(0,len(temp),2)])
                else : print("Input SBCT format illegal !\n")
            elif whichOne == "2" :
                temp = input(counterHelpStr[:-1]+" : ")
                if checkAll(temp, 4) :
                    argumentDict["Counter"] = temp[2:]+temp[0:2]
                else : print("Input Counter format illegal !\n")
            elif whichOne == "3" :
                temp = input(periodHelpStr[:-1]+" : ")
                if checkAll(temp, 4) :
                    argumentDict["Period"] = temp[2:]+temp[0:2]
                else : print("Input Period format illegal !\n")
            else :
                print("!!Not legal input options !!\n")
        
        ModifyWriteBin(WriteDataFileName, argumentDict)

def ModifyWriteBin(WriteDataFileName, argumentDict):
    #PrintWriteInfoSize(WriteDataFileName)
    with open(WriteDataFileName, "rb+") as writeData :
        print("Opened Modify Input-Data name : " + WriteDataFileName )
        #data = writeData.read()
        #print(data)
        for x in argumentDict :
            if argumentDict[x] and x in binAreaStart :
                #print(x)
                #print(argumentDict[x])
                writeData.seek(int(hex(binAreaStart[x]),0),0) 
                for y in range(0,len(argumentDict[x]),2) : # two byte write once
                    #print(argumentDict[x][y]+argumentDict[x][y+1])
                    writeData.write(bytes([int(argumentDict[x][y]+argumentDict[x][y+1],16)]))
        #writeData.seek(int(hex(binAreaLen["startOffset"]),0),0)
        #writeData.write(bytes([int("44",16)]))

#print(int( hex(binAreaLen["startOffset"]),0 ))
#print(bytes([int("9A",16)]))
#print(bytes([136]))
#WriteDataFileName = "write_data.bin"
#ModifyWriteBin(WriteDataFileName, argumentDict)

choce_file.py
import os                                   # operate file
import re                                   # find file

def Choce_file(file_type) : # type : xml / bin
    returnName = None
    fatherDir = os.getcwd()
    allDir = os.listdir(fatherDir)
    if file_type == "bin" :
        fileNameList = re.compile("[^write].*\." + file_type)
    else :
        fileNameList = re.compile(".*\." + file_type)
    fileNameList = list( filter( fileNameList.match, allDir ) )

    if len(fileNameList) >= 2 : 
        print("Finded multy ." + file_type + " : ", fileNameList )
        fileNameList = ["("+str(i+1)+")"+fileNameList[i] for i in range(len(fileNameList))]
        fileNameString = ""
        for i in fileNameList : fileNameString += i + " "
        fileNameString = "Input which NUMBER ." \
                            + file_type \
                            + " you want " \
                            + fileNameString \
                            + "\nor press enter to use " \
                            + fileNameList[0][3:] \
                            + " : "
        fileNameNum = input( fileNameString ) or 1
        while not 0 < int(fileNameNum) <= len(fileNameList) :
            print("Not legal input !!!")
            fileNameNum = input( fileNameString ) or 1
        fileName = fileNameList[int(fileNameNum)-1]
        returnName = fileName[3:]
    elif len(fileNameList) == 1 :
        returnName = fileNameList[0]
    else :
        print("Can not find any ." + file_type + " !!!" )
    
    return returnName

getXMLInfo.py
import xml.etree.ElementTree as ET          # xml

def GetLocation_XMLInfo(tree, targetName, targetAct):
    root = tree.getroot()
    targetValue=[]
    for i in targetName :
        try:
            temp = root.find(i).text
            targetValue.append(temp)
        except Exception as maybeInSomones_AddressOffset :
            finded = False
            for j in root :
                for k in j :
                    if k.tag == "Address" and not finded :
                        #print(k ,"k yes")
                        if k.find("Offset").find("Name").text == i:
                            #print(k.find("Offset").find("Value").text)
                            targetValue.append(k.find("Offset").find("Value").text)
                            finded = True
                    elif not finded : # more deep like PeiA / DxeA
                        for l in k :
                            if l.tag == "Address" :
                                if l.find("Offset").find("Name").text == i :
                                    targetValue.append(l.find("Offset").find("Value").text)
                                    finded = True
                                    break
                    else : # can not find this name
                        print("Can not find this name : " + i )
                        print("Skip it !! ")
                        targetAct.pop(targetName.index(i)-1)
                    if finded :
                        break
                if finded :
                    break
    ## Count location
    targetValueINT=[int(i, 16) for i in targetValue]
    location = targetValueINT[0]
    totalCount = []
    totalCount.append(targetValueINT[0])
    if len(targetValueINT) >= 2 :
        for i in range(1, len(targetValueINT)) :
            if targetAct[i-1] == "+" :
                location += targetValueINT[i]
                totalCount.append(targetAct[i-1])
            elif targetAct[i-1] == "-" :
                location -= targetValueINT[i]
            else :
                print("Don't supoort act : " + targetAct[i-1] + " !! Only +/- !!" )
                print("Skip this act and value : " + targetAct[i-1] + ","+ targetValueINT[i])
                break
    else :
        location = None
        print("Don't have enough number !!")
        print("EXIT !!!")
    return location

modbin_withxml.py
#############################
# UTF-8
# Shawn.Peng@quantatw.com
# Can modify binary with location recorded by xml
############################
# Work flow :
# 1.Can change target.txt to modify more step
#  (1) Default BiosStartInFullRomImage + PcdFlashNvStorageVariableOffset
#  (2) Rule : one line of feature-name and one line of act 
#  (3) The act only suppout +/-
#  (4) If act isn't correct or feature-name cannot find than it will be skiped
# 2.Find feature-name which user set in {Platform}.xml
#  (1) Now can find feature-name in tag of first layer and its text is the value we want
#  (2) Now can find feature-name in value of "name" tag of forth or fifth layers 
#      and the value of the tag "value" of the same layer is what we want
#  (3) User can choice number or press enter to use .xml witch first finded
# 3.Count the value and print it to double check
# 4.Read write_data.bin which will replace 
#  (1) User can input manually to modify when execution CT/MAC/ForceNetBoot/counter/timer
#      otherwise default 11002200...EE00/FF..FF/1/5/2 
#  (2) User can also modify by command line use parameters
#  (3) Format for 4-1: \x00\x00\..\x07 in one line
#  (4) Format for 4-2: number of non-spaced strings from 0-9 and A-F
#  (5) If use argument -y it will automatically confirm.
#  (6) Please note that the .bin file will be modified if the argument y and 
#      the four modification arguments m, s, c, and p are used together
# 5.Write date to {Platform_version}.bin
#  (1)Copy {Platform_version}.bin and rename {Platform_version}_ori.bin
#   progrm will modify in {Platform_version}.bin
# 
# 
# Share point
# xxx\CMIT_BIOS\Tools\modbin_withxml(M1Inject)\modBIN_withXML_7z_v{before v3}.7z
# xxx\CMIT_BIOS\Tools\modbin_withxml(M1Inject)\modBIN_withXML(M1Inject)_v4.7z
# GitHub 
# https://github.com/yishawnpeng/modBIN_withXML
#
#############################
# In write_data_default.bin :
#         CT start at : 204
#            end   at : 234
#         MAC start at : 234
#             end   at : 240
#         BootCounter start at : 243
#                     end   at : 245
#         Period start at : 245
#                end   at : 247
#############################
# NOTICE!! : in xml, <PCD> Block Skip
# NOTICE!! : Using -y and [-m,-s,-c,-p] together will modify the .bin
#############################
### Get and Count location
#import xml.etree.ElementTree as ET          # xml
import shutil                               # copy file
import sys                                  # exit
#import os                                   # operate file
#import re                                   # find file
from getXMLInfo import *                   # will import ET
from choceFile import *                    # will import os/re
from argParse import *
from modifyBin import *

#argumentDict={"MAC":None,"SBCT":None,"Counter":None,"Period":None} ## defined in argParse.py
autoUpdate, legalParmeter = argParse_contant(argumentDict)
if not legalParmeter : sys.exit()

# That user chose .xml
XMLFileName = ChoceFile("xml",autoUpdate)
if not XMLFileName : sys.exit()
#print(XMLFileName)

## Get target
TargetFileName = "target.txt"
targetName =[]
targetAct=[]
with open(TargetFileName, "r") as targetFile :
    print("Opened Target name : " + TargetFileName )
    for i in targetFile.readlines() :
        if i[0:3] != "act":
            #maybe at end don't have \n
            targetName.append(i[:-1] if i[-1]=="\n" else i)
        else :
            targetAct.append(i[-2])       
## Get xml info
tree = ET.parse(XMLFileName)
print("Opened XML name : " + XMLFileName )
location = GetLocation_XMLInfo(tree, targetName, targetAct)
if not location : sys.exit()

#print(targetName)
#print(targetValue)
#print(targetValueINT)
#print(targetAct)
print("Taget location at :" , hex(location))

### Mod bin
# Always use write_data_default.bin to refer
onlyArg=False
WriteDataFileName = "write_data_default.bin"
if not autoUpdate :                                                             # no -y
    if list(argumentDict.values()).count(None) != len(argumentDict.keys()) :    # only arg 
        onlyArg=True
        BinaryFileName = ChoceFile("bin", autoUpdate)
        if not BinaryFileName : sys.exit()

        with open(BinaryFileName, "rb+") as binaryFile :
            binaryFile.seek(location)
            data = binaryFile.read(sum(list(binAreaLen.values())))

        tmepName = BinaryFileName[:-4]+"_temp"+BinaryFileName[-4:]
        shutil.copy2(BinaryFileName, tmepName)
        with open(tmepName, "wb+") as binaryFile :
            binaryFile.write(data)

        ModifyWriteBin(tmepName, argumentDict)
        WriteDataFileName = tmepName
    else :                                                                      # all arg empty -> manul
        WriteDataFileName = ChoceFile(WriteDataFileName, autoUpdate)
        if not WriteDataFileName : sys.exit()
        # If noy -y let user modify manuly
        ModifyWriteBin_manul(WriteDataFileName, argumentDict)
else :  # is autoUpdate                                                         # have y -> write all data 
    if list(argumentDict.values()).count(None) != len(argumentDict.keys()) :    # mod some in bin
        ModifyWriteBin(WriteDataFileName, argumentDict)
    else :  pass                                                                # only -y 
    
with open(WriteDataFileName, "rb") as writeData :
    print("Opened Input-Data name : " + WriteDataFileName )
    data = writeData.read()

if not onlyArg :                # only arg -> already chose
    # That user chose {written}.bin
    BinaryFileName = ChoceFile("bin", autoUpdate)
    if not BinaryFileName : sys.exit()

#copy _ori.bin and write data to .bin
shutil.copy2(BinaryFileName, BinaryFileName[:-4]+"_ori"+BinaryFileName[-4:])
print("Created _ori.bin ")
with open(BinaryFileName, "rb+") as binaryFile :
    print("Opened Binary name : " + BinaryFileName )
    binaryFile.seek(location)
    binaryFile.write(data)

if onlyArg :                    # only arg
    if os.path.exists(tmepName):
        os.remove(tmepName)

print("Write Complete !")
